### PR TITLE
Feat/question table

### PR DESCRIPTION
**Pull request by lustigy96**

### CODE REVIEW

File: shared/question/question.ts [hunk 1]
```diff
@@ -6,12 +6,20 @@ interface ScaleLimit {
...
+    required: boolean;
+}
+
 enum QuestionType {
     multipleChoice = 'MULTIPLE_CHOICE',
     open = 'OPEN',
     checkbox = 'CHECKBOX',
     linearScale = 'LINEAR_SCALE',
     date = 'DATE',
+    multipleChoiseTable = 'MULTIPLE_CHOISE_TABLE',
```
Comment (carmeltsadok) [inline shared/question/question.ts:20]: Change to multiple **choice** table

**Review by carmeltsadok — CHANGES_REQUESTED** [review]

File: shared/question/question.ts [hunk 2]
```diff
@@ -181,6 +194,99 @@ class DateQuestion extends Question {
...
+    isAnswerValid() {
+        if (this.answer.length !== this.rows.length) return false;
+        const answerValidation = this.rows.map((row, idx) => {
+            if (this.answer[idx] < 0 || this.answer[idx] > this.columns.length - 1 || (row.required && this.answer[idx] === undefined)) {
+                return false;
+            }
+            return true;
+        });
+        return !answerValidation.includes(false) && super.isAnswerValid();
+    }
```
Comment (PCOffline) [inline shared/question/question.ts:224]: Use `Array.prototype.every` instead. It's a significantly faster method.

```suggestion
        return this.rows.every((row, idx) => this.answer[idx] > 0 && this.answer[idx] < this.columns.length  && (!row.required || this.answer[idx])) && super.isAnswerValid();
    }
```

File: shared/question/question.ts [hunk 3]
```diff
@@ -181,6 +194,99 @@ class DateQuestion extends Question {
...
+}
+
+class CheckboxTableQuestion extends Question {
+    /**
+     * table is from one of the following types: MULTIPLE_CHOICE_TABLE, CHECKBOX_TABLE, TEXT_TABLE
+     * CHECKBOX_TABLE: each row represents a checkbox question
+     * answer is a 2-dimetions array, size: (rows x ?), each sub array includes the indexes of the picked column, for each row accordingly     *
```
Comment (PCOffline) [inline shared/question/question.ts:231]: 2-dimensional array*

File: shared/question/question.ts [hunk 4]
```diff
@@ -181,6 +194,99 @@ class DateQuestion extends Question {
...
+
+    constructor(questionObject: IQuestion) {
+        super(questionObject);
+        this.columns = questionObject.columns;
+        this.rows = questionObject.rows;
+    }
+
+    isAnswerValid() {
+        if (this.answer.length !== this.rows.length) return false;
+        const answerValidation = this.rows.map((row, idx) => {
```
Comment (PCOffline) [inline shared/question/question.ts:247]: Again, use `Array.prototype.every` instead.

File: shared/question/question.ts [hunk 5]
```diff
@@ -181,6 +194,99 @@ class DateQuestion extends Question {
...
```
Comment (PCOffline) [inline shared/question/question.ts:277]: Use `Array.prototype.every` instead.

**Review by PCOffline — CHANGES_REQUESTED** [review]

File: shared/question/question.ts [hunk 6]
```diff
@@ -181,6 +194,91 @@ class DateQuestion extends Question {
...
+        this.columns = questionObject.columns;
+        this.rows = questionObject.rows;
+    }
+
+    isAnswerValid() {
+        if (this.answer.length !== this.rows.length) return false;
+        const answerValidation = this.rows.every(
+            (row, idx) =>
+                this.answer[idx].length === this.columns.length && // check fill length of each row
+                (!this.required || !(this.answer[idx].includes('') || !this.answer[idx].includes(undefined))), // if answer is required, validate no undefineds ot empty strings
```
Comment (PCOffline) [inline shared/question/question.ts:276]: `!(this.answer[idx].includes('') || !this.answer[idx].includes(undefined))` will be true if the answer includes undefined.
Also, `String.includes(undefined)` will be true only if it includes the string `'undefined'`, is this the behaviour you expect?

**Review by carmeltsadok — APPROVED** [review]

File: shared/question/question.ts (hunk 6 above)
Comment (PCOffline) [reply shared/question/question.ts:276]: This is still unfixed

Comment (carmeltsadok) [reply shared/question/question.ts:276]: @lustigy96

Comment (lustigy96) [reply shared/question/question.ts:276]: this.answer[idx] is an array of strings and not a string.

Comment (PCOffline) [reply shared/question/question.ts:276]: Okay

**Review by PCOffline — APPROVED** [review]

**Pull request merged**